ssl/d1_lib.c

```diff
@@ -246,16 +251,26 @@ void dtls1_start_timer(SSL *s)
     }
 #endif
 
-    /* If timer is not set, initialize duration with 1 second */
+    /* If timer is not set, initialize duration with 1 second or
+     * a user-specified value if the timer callback is installed. */
     if (s->d1->next_timeout.tv_sec == 0 && s->d1->next_timeout.tv_usec == 0) {
-        s->d1->timeout_duration = 1;
+
+        if (s->d1->timer_cb != NULL)
+            s->d1->timeout_duration_ms = s->d1->timer_cb(s, 1000);
+        else
+            s->d1->timeout_duration_ms = 1000;
     }
 
     /* Set timeout to current time */
     get_current_time(&(s->d1->next_timeout));
 
     /* Add duration to current time */
-    s->d1->next_timeout.tv_sec += s->d1->timeout_duration;
+    s->d1->next_timeout.tv_usec += s->d1->timeout_duration_ms * 1000;
+    if (s->d1->next_timeout.tv_usec >= 1000000) {
+        s->d1->next_timeout.tv_sec++;
+        s->d1->next_timeout.tv_usec -= 1000000;
+    }
```

Comment:
If a call to `s->d1->timer_cb()` set `s->d1->timeout_duration_ms` to, say, 3000, this operation will not be quite right...  If you change `if` to `while`, that would be a cheap fix ;-)